edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs

```diff
@@ -0,0 +1,191 @@
+        }
+
+        public async Task SyncToCloud(string id)
+        {
+            Events.SyncingReportedPropertiesToCloud(id);
+            Option<TwinStoreEntity> twinWithReportedProperties = (await this.twinStore.Get(id))
+                .Filter(t => t.ReportedPropertiesPatch.HasValue);
+            if (twinWithReportedProperties.HasValue)
+            {
+                using (await this.lockProvider.GetLock(id).LockAsync())
+                {
+                    Events.StoredReportedPropertiesFound(id);
+                    Option<TwinStoreEntity> twinInfo = await this.twinStore.Get(id);
+                    await twinInfo.ForEachAsync(
+                        async ti =>
+                        {
+                            await ti.ReportedPropertiesPatch.ForEachAsync(
+                                async reportedPropertiesPatch =>
+                                {
+                                    bool result = await this.cloudSync.UpdateReportedProperties(id, reportedPropertiesPatch);
+                                    if (result)
+                                    {
+                                        Events.UpdateReportedPropertiesSucceeded(id);
+                                        await this.twinStore.Update(
+                                            id,
+                                            t => new TwinStoreEntity(t.Twin, Option.None<TwinCollection>()));
+                                        twinInfo = await this.twinStore.Get(id);
+                                    }
+                                    else
+                                    {
+                                        Events.UpdateReportedPropertiesFailed(id);
+                                    }
+                                });
+                        });
+                }
+            }
+            else
+            {
+                Events.DoneSyncingReportedProperties(id);
+            }
+        }
+
+        async Task SyncToCloud()
+        {
+            while (true)
+            {
+                try
+                {
+                    // Take a snapshot of clients to process
+                    IEnumerable<string> clientsToProcess;
+                    lock (this.syncToCloudSetLock)
+                    {
+                        clientsToProcess = this.syncToCloudClients.ToList();
+                    }
+
+                    foreach (string id in clientsToProcess)
+                    {
+                        await this.SyncToCloud(id);
+                    }
+                }
+                catch (Exception e)
+                {
+                    Events.ErrorSyncingReportedPropertiesToCloud(e);
+                }
+
+                // Wait for syncfrequency to avoid looping too fast,
+                // then wait for the signal indicating more work is ready
+                await Task.Delay(this.syncFrequency);
+                await this.syncToCloudSignal.WaitAsync();
```

Comment:
Who sets this signal? I see it in the InitSync call. Anywhere else?

Reply:
Ah. It's set externally.

Reply:
Right, every time a client updates its reported properties, we signal this loop to get going again.